ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -620,42 +694,38 @@ modifyState VolatileDBEnv{_dbHasFS = hasFS :: HasFS m h, ..} action = do
     ErrorHandling{..} = _dbErr
     HasFS{..}         = hasFS
 
-    open :: m (OpenOrClosed blockId h)
+    open :: m (OpenOrClosed blockId h m)
     open = takeMVar _dbInternalState
 
     close
-      :: OpenOrClosed blockId h
-      -> ExitCase (Either (VolatileDBError blockId) (InternalState blockId h, r))
+      :: OpenOrClosed blockId h m
+      -> ExitCase (Either (VolatileDBError blockId) (InternalState blockId m h, r))
       -> m ()
     close mst ec = case ec of
       -- Restore the original state in case of an abort.
       ExitCaseAbort         -> putMVar _dbInternalState mst
       -- In case of an exception, close the DB for safety.
-      ExitCaseException _ex -> do
-        putMVar _dbInternalState VolatileDbClosed
-        closeOpenHandle mst
+      ExitCaseException _ex ->
+        unsafeReleaseAll _registry
+          `finally` putMVar _dbInternalState VolatileDbClosed
```

Comment:
Why not:
```haskell
        putMVar _dbInternalState VolatileDbClosed
        unsafeReleaseAll _registry
```
The `putMVar` should never fail, right?

Reply:
I think it's the same case as this https://github.com/input-output-hk/ouroboros-network/pull/1165#discussion_r345820664
cleaning the registry should happen with the `InternalState` lock.